File: PHP/qr_code_utils/qr_code_utils_test.php
<?php
/**
 * AllToolkit - QR Code Utilities Test Suite
 *
 * Comprehensive test suite for QR Code generation functionality.
 *
 * @package AllToolkit\QrCodeUtils
 */

require_once __DIR__ . '/mod.php';

use AllToolkit\QrCodeGenerator;
use AllToolkit\QrCodeUtils;
use AllToolkit\QrCodeException;

/**
 * Simple test runner
 */
class TestRunner {
    private $tests = 0;
    private $passed = 0;
    private $failed = 0;
    private $errors = [];

    public function assert($condition, $message) {
        $this->tests++;
        if ($condition) {
            $this->passed++;
            echo "  ✓ {$message}\n";
        } else {
            $this->failed++;
            $this->errors[] = $message;
            echo "  ✗ {$message}\n";
        }
    }

    public function assertEquals($expected, $actual, $message) {
        $this->assert($expected === $actual, $message . " (expected: {$expected}, got: {$actual})");
    }

    public function assertTrue($condition, $message) {
        $this->assert($condition, $message);
    }

    public function assertFalse($condition, $message) {
        $this->assert(!$condition, $message);
    }

    public function assertException($callback, $expectedMessage = null, $message = 'Should throw exception') {
        $this->tests++;
        try {
            $callback();
            $this->failed++;
            $this->errors[] = $message;
            echo "  ✗ {$message} (no exception thrown)\n";
        } catch (QrCodeException $e) {
            if ($expectedMessage === null || strpos($e->getMessage(), $expectedMessage) !== false) {
                $this->passed++;
                echo "  ✓ {$message}\n";
            } else {
                $this->failed++;
                $this->errors[] = $message;
                echo "  ✗ {$message} (wrong message: {$e->getMessage()})\n";
            }
        }
    }

    public function summary() {
        echo "\n";
        echo "========================================\n";
        echo "Test Results: {$this->passed}/{$this->tests} passed\n";
        echo "========================================\n";

        if ($this->failed > 0) {
            echo "\nFailed tests:\n";
            foreach ($this->errors as $error) {
                echo "  - {$error}\n";
            }
        }

        return $this->failed === 0;
    }
}

// Run tests
$runner = new TestRunner();

echo "========================================\n";
echo "QR Code Utils Test Suite\n";
echo "========================================\n\n";

// Test 1: Constructor validation
echo "Test Group: Constructor Validation\n";
echo "------------------------------------\n";

try {
    $qr = new QrCodeGenerator(1, QrCodeGenerator::ERROR_CORRECTION_L);
    $runner->assertTrue($qr instanceof QrCodeGenerator, "Create QR Code generator v1, level L");
} catch (Exception $e) {
    $runner->assertFalse(true, "Create QR Code generator v1, level L: " . $e->getMessage());
}

$runner->assertException(function() {
    new QrCodeGenerator(0, QrCodeGenerator::ERROR_CORRECTION_L);
}, "Version must be between", "Invalid version (0) should throw exception");

$runner->assertException(function() {
    new QrCodeGenerator(11, QrCodeGenerator::ERROR_CORRECTION_L);
}, "Version must be between", "Invalid version (11) should throw exception");

$runner->assertException(function() {
    new QrCodeGenerator(1, -1);
}, "Invalid error correction level", "Invalid error correction level (-1) should throw exception");

$runner->assertException(function() {
    new QrCodeGenerator(1, 4);
}, "Invalid error correction level", "Invalid error correction level (4) should throw exception");

echo "\n";

// Test 2: Mode detection
echo "Test Group: Mode Detection\n";
echo "------------------------------------\n";

$runner->assertEquals(QrCodeGenerator::MODE_NUMERIC, QrCodeGenerator::detectMode("1234567890"), "Detect numeric mode");
$runner->assertEquals(QrCodeGenerator::MODE_NUMERIC, QrCodeGenerator::detectMode("0"), "Detect numeric mode (single digit)");
$runner->assertEquals(QrCodeGenerator::MODE_ALPHANUMERIC, QrCodeGenerator::detectMode("HELLO WORLD"), "Detect alphanumeric mode");
$runner->assertEquals(QrCodeGenerator::MODE_ALPHANUMERIC, QrCodeGenerator::detectMode("ABC123"), "Detect alphanumeric mode (mixed)");
$runner->assertEquals(QrCodeGenerator::MODE_BYTE, QrCodeGenerator::detectMode("Hello World"), "Detect byte mode (lowercase)");
$runner->assertEquals(QrCodeGenerator::MODE_BYTE, QrCodeGenerator::detectMode("Hello, World!"), "Detect byte mode (punctuation)");
$runner->assertEquals(QrCodeGenerator::MODE_BYTE, QrCodeGenerator::detectMode("你好世界"), "Detect byte mode (Unicode)");

echo "\n";

// Test 3: QR Code generation
echo "Test Group: QR Code Generation\n";
echo "------------------------------------\n";

$qr = new QrCodeGenerator(1, QrCodeGenerator::ERROR_CORRECTION_L);
$modules = $qr->generate("HELLO");

$runner->assertTrue(is_array($modules), "Generate returns array");
$runner->assertEquals(21, count($modules), "Version 1 has 21x21 modules");
$runner->assertEquals(21, count($modules[0]), "Each row has 21 modules");
$runner->assertEquals(1, $qr->getVersion(), "Get version returns 1");
$runner->assertEquals(21, $qr->getModuleCount(), "Get module count returns 21");
$runner->assertEquals('L', $qr->getErrorCorrectionLevelName(), "Get error correction level name returns L");

echo "\n";

// Test 4: ASCII output
echo "Test Group: ASCII Output\n";
echo "------------------------------------\n";

$qr = new QrCodeGenerator(1, QrCodeGenerator::ERROR_CORRECTION_L);
$ascii = $qr->toAscii("TEST");

$runner->assertTrue(strpos($ascii, '██') !== false, "ASCII output contains dark modules");
$runner->assertTrue(strpos($ascii, '  ') !== false, "ASCII output contains light modules");
$runner->assertTrue(strpos($ascii, "\n") !== false, "ASCII output contains newlines");

echo "\n";

// Test 5: SVG output
echo "Test Group: SVG Output\n";
echo "------------------------------------\n";

$qr = new QrCodeGenerator(1, QrCodeGenerator::ERROR_CORRECTION_L);
$svg = $qr->toSvg("TEST");

$runner->assertTrue(strpos($svg, '<?xml') !== false, "SVG starts with XML declaration");
$runner->assertTrue(strpos($svg, '<svg') !== false, "SVG contains svg element");
$runner->assertTrue(strpos($svg, '</svg>') !== false, "SVG ends with closing svg tag");
$runner->assertTrue(strpos($svg, 'rect') !== false, "SVG contains rect elements");
$runner->assertTrue(strpos($svg, 'width="') !== false, "SVG has width attribute");
$runner->assertTrue(strpos($svg, 'height="') !== false, "SVG has height attribute");

echo "\n";

// Test 6: Matrix output
echo "Test Group: Matrix Output\n";
echo "------------------------------------\n";

$qr = new QrCodeGenerator(1, QrCodeGenerator::ERROR_CORRECTION_L);
$matrix = $qr->toMatrix("TEST");

$runner->assertTrue(is_array($matrix), "Matrix is array");
$runner->assertEquals(21, count($matrix), "Matrix has 21 rows");
$runner->assertEquals(21, count($matrix[0]), "Matrix has 21 columns");

// Check that matrix contains only 0s and 1s
$validValues = true;
for ($i = 0; $i < 21; $i++) {
    for ($j = 0; $j < 21; $j++) {
        if ($matrix[$i][$j] !== 0 && $matrix[$i][$j] !== 1) {
            $validValues = false;
            break 2;
        }
    }
}
$runner->assertTrue($validValues, "Matrix contains only 0s and 1s");

echo "\n";

// Test 7: Utility functions
echo "Test Group: Utility Functions\n";
echo "------------------------------------\n";

$ascii = QrCodeUtils::toAscii("HELLO");
$runner->assertTrue(strpos($ascii, '██') !== false, "QrCodeUtils::toAscii generates output");

$svg = QrCodeUtils::toSvg("WORLD");
$runner->assertTrue(strpos($svg, '<svg') !== false, "QrCodeUtils::toSvg generates SVG");

$matrix = QrCodeUtils::toMatrix("TEST123");
$runner->assertTrue(is_array($matrix), "QrCodeUtils::toMatrix returns array");

$runner->assertEquals('Numeric', QrCodeUtils::getModeName(QrCodeGenerator::MODE_NUMERIC), "Get mode name for numeric");
$runner->assertEquals('Alphanumeric', QrCodeUtils::getModeName(QrCodeGenerator::MODE_ALPHANUMERIC), "Get mode name for alphanumeric");
$runner->assertEquals('Byte', QrCodeUtils::getModeName(QrCodeGenerator::MODE_BYTE), "Get mode name for byte");

$runner->assertEquals('L', QrCodeUtils::getErrorCorrectionLevelName(QrCodeGenerator::ERROR_CORRECTION_L), "Get EC level name L");
$runner->assertEquals('M', QrCodeUtils::getErrorCorrectionLevelName(QrCodeGenerator::ERROR_CORRECTION_M), "Get EC level name M");
$runner->assertEquals('Q', QrCodeUtils::getErrorCorrectionLevelName(QrCodeGenerator::ERROR_CORRECTION_Q), "Get EC level name Q");
$runner->assertEquals('H', QrCodeUtils::getErrorCorrectionLevelName(QrCodeGenerator::ERROR_CORRECTION_H), "Get EC level name H");

echo "\n";

// Test 8: Different versions
echo "Test Group: Different Versions\n";
echo "------------------------------------\n";

for ($v = 1; $v <= 5; $v++) {
    $qr = new QrCodeGenerator($v, QrCodeGenerator::ERROR_CORRECTION_L);
    $modules = $qr->generate("TEST");
    $expectedSize = 17 + 4 * $v;
    $runner->assertEquals($expectedSize, count($modules), "Version {$v} has {$expectedSize}x{$expectedSize} modules");
}

echo "\n";

// Test 9: Different error correction levels
echo "Test Group: Error Correction Levels\n";
echo "------------------------------------\n";

$levels = [
    QrCodeGenerator::ERROR_CORRECTION_L => 'L',
    QrCodeGenerator::ERROR_CORRECTION_M => 'M',
    QrCodeGenerator::ERROR_CORRECTION_Q => 'Q',
    QrCodeGenerator::ERROR_CORRECTION_H => 'H',
];

foreach ($levels as $level => $name) {
    $qr = new QrCodeGenerator(1, $level);
    $modules = $qr->generate("HELLO");
    $runner->assertTrue(is_array($modules), "Generate with error correction level {$name}");
    $runner->assertEquals($name, $qr->getErrorCorrectionLevelName(), "Get EC level name {$name}");
}

echo "\n";

// Test 10: Various data types
echo "Test Group: Various Data Types\n";
echo "------------------------------------\n";

$testData = [
    "1234567890" => "Numeric data",
    "HELLO WORLD" => "Alphanumeric data",
    "Hello, World!" => "Byte data with punctuation",
    "test@example.com" => "Email address",
    "https://example.com" => "URL",
    "Test123" => "Mixed alphanumeric",
];

foreach ($testData as $data => $description) {
    $qr = new QrCodeGenerator(2, QrCodeGenerator::ERROR_CORRECTION_L);
    $modules = $qr->generate($data);
    $runner->assertTrue(is_array($modules), "Generate QR for {$description}");
}

echo "\n";

// Test 11: Custom ASCII characters
echo "Test Group: Custom ASCII Characters\n";
echo "------------------------------------\n";

$qr = new QrCodeGenerator(1, QrCodeGenerator::ERROR_CORRECTION_L);
$ascii = $qr->toAscii("TEST", "##", "..");
$runner->assertTrue(strpos($ascii, '##') !== false, "Custom dark character works");
$runner->assertTrue(strpos($ascii, '..') !== false, "Custom light character works");

echo "\n";

// Test 12: Custom SVG colors
echo "Test Group: Custom SVG Colors\n";
echo "------------------------------------\n";

$qr = new QrCodeGenerator(1, QrCodeGenerator::ERROR_CORRECTION_L);
$svg = $qr->toSvg("TEST", 4, '#ff0000', '#ffffff');
$runner->assertTrue(strpos($svg, '#ff0000') !== false, "Custom dark color in SVG");
$runner->assertTrue(strpos($svg, '#ffffff') !== false, "Custom light color in SVG");

echo "\n";

// Print summary
$success = $runner->summary();

exit($success ? 0 : 1);

File: Ruby/qr_code_utils/qr_code_utils_test.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# QR Code Utilities Test Suite
# Comprehensive tests for QR code generation

require_relative 'mod'

class QrCodeUtilsTest
  include AllToolkit::QrCodeUtils

  def self.run_all
    puts "Running QR Code Utils Tests..."
    puts "=" * 50

    test = new
    methods = public_methods(false).select { |m| m.to_s.start_with?('test_') }
    passed = 0
    failed = 0

    methods.each do |method|
      begin
        test.send(method)
        puts "✓ #{method}"
        passed += 1
      rescue => e
        puts "✗ #{method}: #{e.message}"
        failed += 1
      end
    end

    puts "=" * 50
    puts "Results: #{passed} passed, #{failed} failed"
    exit(failed > 0 ? 1 : 0)
  end

  def assert_equal(expected, actual, message = nil)
    return if expected == actual
    raise "#{message || 'Assertion failed'}: expected #{expected.inspect}, got #{actual.inspect}"
  end

  def assert(condition, message = nil)
    return if condition
    raise message || 'Assertion failed'
  end

  def assert_raises(error_class)
    begin
      yield
      raise "Expected #{error_class} to be raised"
    rescue error_class
      # Expected
    end
  end

  # Test basic QR code generation
  def test_basic_generation
    qr = QrCodeUtils.generate("Hello")
    assert_equal 21, qr.size # Version 1 = 21x21
    assert_equal "Hello", qr.data
    assert_equal 1, qr.version
  end

  # Test different error correction levels
  def test_error_correction_levels
    qr_l = QrCodeUtils.generate("Test", QrCodeUtils::LEVEL_L)
    qr_m = QrCodeUtils.generate("Test", QrCodeUtils::LEVEL_M)
    qr_q = QrCodeUtils.generate("Test", QrCodeUtils::LEVEL_Q)
    qr_h = QrCodeUtils.generate("Test", QrCodeUtils::LEVEL_H)

    assert_equal 0, qr_l.error_correction_level
    assert_equal 1, qr_m.error_correction_level
    assert_equal 2, qr_q.error_correction_level
    assert_equal 3, qr_h.error_correction_level
  end

  # Test version detection for different data lengths
  def test_version_detection
    qr1 = QrCodeUtils.generate("A" * 10)  # Version 1
    qr2 = QrCodeUtils.generate("A" * 30)  # Version 2
    qr3 = QrCodeUtils.generate("A" * 50)  # Version 3

    assert_equal 1, qr1.version
    assert_equal 2, qr2.version
    assert_equal 3, qr3.version
  end

  # Test ASCII output
  def test_to_ascii
    qr = QrCodeUtils.generate("Hi")
    ascii = qr.to_ascii
    lines = ascii.split("\n")

    # Should have border + QR size + border
    assert_equal 21 + 4, lines.length # 2 border each side
    assert lines.all? { |l| l.length == 21 + 4 } # All lines same width
  end

  # Test Unicode output
  def test_to_unicode
    qr = QrCodeUtils.generate("Hi")
    unicode = qr.to_unicode
    lines = unicode.split("\n")

    assert lines.length > 0
    assert lines.all? { |l| l.length > 0 }
  end

  # Test SVG output
  def test_to_svg
    qr = QrCodeUtils.generate("Test")
    svg = qr.to_svg

    assert svg.include?('<?xml version="1.0"')
    assert svg.include?('<svg')
    assert svg.include?('</svg>')
    assert svg.include?('<rect')
  end

  # Test matrix output
  def test_to_matrix
    qr = QrCodeUtils.generate("Test")
    matrix = qr.to_matrix

    assert_equal qr.size, matrix.length
    assert matrix.all? { |row| row.length == qr.size }
  end

  # Test module access
  def test_module_access
    qr = QrCodeUtils.generate("Test")

    # Top-left finder pattern should have dark modules
    assert qr[0, 0] # Corner should be dark
    assert qr[6, 6] # Center should be dark
    assert !qr[1, 1] # Inner should be light
  end

  # Test data too long raises error
  def test_data_too_long
    assert_raises(ArgumentError) do
      QrCodeUtils.generate("A" * 300) # Too long for version 10
    end
  end

  # Test empty string
  def test_empty_string
    qr = QrCodeUtils.generate("")
    assert_equal "", qr.data
    assert_equal 1, qr.version
  end

  # Test special characters
  def test_special_characters
    qr = QrCodeUtils.generate("Hello, World! @#$%")
    assert_equal "Hello, World! @#$%", qr.data
  end

  # Test Unicode characters
  def test_unicode_characters
    qr = QrCodeUtils.generate("你好世界")
    assert_equal "你好世界", qr.data
  end

  # Test URL generation
  def test_url_generation
    url = "https://example.com/path?query=value"
    qr = QrCodeUtils.generate(url)
    assert_equal url, qr.data
  end

  # Test max data length helper
  def test_max_data_length
    assert QrCodeUtils.max_data_length(1, QrCodeUtils::LEVEL_L) > 0
    assert QrCodeUtils.max_data_length(10, QrCodeUtils::LEVEL_H) > 0
  end

  # Test compact ASCII alias
  def test_compact_ascii_alias
    qr = QrCodeUtils.generate("Test")
    assert_equal qr.to_unicode, qr.to_compact_ascii
  end

  # Test different border sizes
  def test_different_border_sizes
    qr = QrCodeUtils.generate("Test")
    ascii1 = qr.to_ascii(1)
    ascii2 = qr.to_ascii(4)

    assert ascii1.lines.count < ascii2.lines.count
  end

  # Test finder patterns are present
  def test_finder_patterns_present
    qr = QrCodeUtils.generate("Test")

    # Top-left finder pattern
    assert qr[0, 0]
    assert qr[0, 6]
    assert qr[6, 0]
    assert qr[6, 6]

    # Top-right finder pattern
    assert qr[0, qr.size - 1]
    assert qr[0, qr.size - 7]
    assert qr[6, qr.size - 1]
    assert qr[6, qr.size - 7]

    # Bottom-left finder pattern
    assert qr[qr.size - 1, 0]
    assert qr[qr.size - 7, 0]
    assert qr[qr.size - 1, 6]
    assert qr[qr.size - 7, 6]
  end

  # Test timing patterns
  def test_timing_patterns
    qr = QrCodeUtils.generate("Test")

    # Timing pattern should alternate
    (8...qr.size - 8).each do |i|
      expected = i.even?
      assert_equal expected, qr[6, i], "Timing pattern at row 6, col #{i}"
      assert_equal expected, qr[i, 6], "Timing pattern at row #{i}, col 6"
    end
  end

  # Test dark module
  def test_dark_module
    qr = QrCodeUtils.generate("Test")
    # Dark module is always at (4*version + 9, 8)
    assert qr[4 * qr.version + 9, 8]
  end

  # Test version 2+ has alignment patterns
  def test_alignment_patterns
    qr1 = QrCodeUtils.generate("A") # Version 1 - no alignment
    qr2 = QrCodeUtils.generate("A" * 25) # Version 2 - has alignment

    # Version 1 should not have alignment patterns in the middle
    # Version 2+ should have them
    assert_equal 1, qr1.version
    assert qr2.version >= 2
  end

  # Test SVG dimensions
  def test_svg_dimensions
    qr = QrCodeUtils.generate("Test")
    svg = qr.to_svg(10, 4)

    # Should contain width and height
    assert svg.include?('width="290"') # (21 + 8) * 10 = 290
    assert svg.include?('height="290"')
  end

  # Test function pattern detection
  def test_function_pattern_detection
    qr = QrCodeUtils.generate("Test")

    # Finder patterns should be marked as function
    assert qr.function?(0, 0)
    assert qr.function?(3, 3)
    assert !qr.function?(10, 10) # Data area
  end

  # Test numeric data
  def test_numeric_data
    qr = QrCodeUtils.generate("1234567890")
    assert_equal "1234567890", qr.data
  end

  # Test long data requires higher version
  def test_long_data_higher_version
    short_qr = QrCodeUtils.generate("Short")
    long_qr = QrCodeUtils.generate("A" * 100)

    assert short_qr.version < long_qr.version
  end

  # Test QR code is square
  def test_qr_code_is_square
    qr = QrCodeUtils.generate("Test data here")
    assert_equal qr.size, qr.modules.length
    assert qr.modules.all? { |row| row.length == qr.size }
  end

  # Test multiple generations produce consistent results
  def test_consistent_generation
    qr1 = QrCodeUtils.generate("Same data")
    qr2 = QrCodeUtils.generate("Same data")
    
    # Same data should produce same QR code
    assert_equal qr1.size, qr2.size
    assert_equal qr1.version, qr2.version
  end
end

# Run tests if executed directly
if __FILE__ == $0
  QrCodeUtilsTest.run_all
end
